### DISCUSSION TITLE
lazyload events?

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#8232] [Usage Questions]
Are there any events or methodologies that would allow users to delay (as in, put behind a mutex) when lazy loads are acted upon by the owning session?
I've looked but Instance.load() etc but this happens late, and other events I've found don't seem appropriate.

If this is not possible, the below details may motivate the feature.

It's worth noting that our prior conversations about possibly unorthodox usage of Session are largely moot: I've dumped all that and taken the plunge to completely change the SQLA usage patterns in that problematic app. Thank goodness.

-------
details

The motivation is to delay resolution of lazy loads until objects that would satisfy the resolution but which are currently detached, are attached.

Irrespective of cleaner session usage there remains what seems to me to be a set of (unfortunately) mutually exclusive choices that must be made when using SQLA in an async environment, and that such usage could be greatly enhanced by a little more sugar in the lazy loading area.  Of course, all the necessary features may already exist (above questions).

Basically it's as follows:
* if you want one proxy per database row, i.e., "one object one graph of those objects"
* and you want lazy loading in general (not purely during a loading phase)
it seems to logically follow that:
* you need a long lived session in which the object graph is add()ed (objects are attached)

It then follows that an arbitrary relationship traversal at an arbitrary time could cause lazy loading, and if any part of that object graph has temporarily been checked out of the long lived session (say, to persist it via a writer session), your loads may happen inconsistent to what actually exists somewhere in memory.

By the way, this exercise allowed me to find a pernicious reason for some of my prior problems:  If, when creating a new object, it has a relationship and that relationship attr is not set to `None` , then traversal will trigger lazy loading ... a missing such initialization was the root cause of some of the database accesses I was trying to carefully control.

My current implementation uses a DBAPI autocommit Session autobegin long lived session for the object graph (call this qtx), and listens for before_flush and after_flush_post_exec events. It uses those to acquire and release a mutex. Meanwhile, objects undergoing creations and modifications are collected and those collections are serialized onto a write worker greenlet when ready. The collected objects are add()ed to the qtx post commit, while holding the same mutex.

This effectively controls writes onto transaction boundaries. Naively commit()ing a common session would otherwise accidentally write incomplete graphs, and for similar reasons, commit()ing directly via the qtx() would not work.

I considered that certain in flight modifications may be being committed by the qtx during lazy-load resolution, but since it's autobegin and I never call commit() I'm thinking that's avoided. Likewise, there may be some lazy loads that are being resolved while certain objects are in a write session, but since their sub-graph is coincident, that's ok.

A.

**Comment by zzzeek** [comment]
> Are there any events or methodologies that would allow users to delay (as in, put behind a mutex) when lazy loads are acted upon by the owning session? I've looked but Instance.load() etc but this happens late, and other events I've found don't seem appropriate.

normally I would have little idea what this means but it seems you have an app where any number of threads are accessing **the same ** pool of objects all at once, which is very much completely the opposite of the assumptions made by the SQLAlchemy ORM, which is very explicit that sessions and the objects are attached to them are completely not threadsafe.   So if you were to have this architecture, I would strongly recommend mutexing it at a coarse-grained level of each app operation, not at the attribute access level which is a really bad idea.

that said, the API to put a mutex around lazy loads is to use [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=do_orm_execute#sqlalchemy.orm.SessionEvents.do_orm_execute) in conjunction with [invoke_statement](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.invoke_statement) to control the statement execution.   

> 
> Irrespective of cleaner session usage there remains what seems to me to be a set of (unfortunately) mutually exclusive choices that must be made when using SQLA in an async environment, 

IIRC you are using gevent, which is not "explicit" async from a Python perspective.   applications that use gevent have no inherent issues using lazy loading, and there are thousands of production applications use SQLAlchemy + gevent + lazy loading without restriction.  but they dont use one global object graph, because all the documentation stresses that this is not the correct way to use the ORM.

> 
> Basically it's as follows:
> 
>     * if you want one proxy per database row, i.e., "one object one graph of those objects"

What you are doing sounds like you have a second-level cache of some kind that's global; when you use that pattern, the correct way to work with some objects such that you are going to read and manipulate them, you have to [merge](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=merge#sqlalchemy.orm.Session.merge) those objects into a local session.     assuming the objects have no changes on them, you can use the "load=False" parameter to avoid actually emitting any SQL.  There's also ways to take objects from that graph, create copies of them, and associate them with any ad-hoc Session where they will then behave as though they were loaded from the database like [make_transient_to_detached](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=detached#sqlalchemy.orm.make_transient_to_detached), basically you can implement home-grown "merge" methods of all kinds to avoid any and all SQL if your "global" state is what you want.    But the key thing with all these methods is, **you must make local copies of objects before working with them**.   


  
> 
> It then follows that an arbitrary relationship traversal at an arbitrary time could cause lazy loading, and if any part of that object graph has temporarily been checked out of the long lived session (say, to persist it via a writer session), your loads may happen inconsistent to what actually exists somewhere in memory.

my recommendation is you would need to ensure you have thread-local, non-global-session local copies of objects at all times.


> 
> My current implementation uses a DBAPI autocommit Session autobegin long lived session for the object graph (call this qtx), and listens for before_flush and after_flush_post_exec events. It uses those to acquire and release a mutex. Meanwhile, objects undergoing creations and modifications are collected and those collections are serialized onto a write worker greenlet when ready. The collected objects are add()ed to the qtx post commit, while holding the same mutex.
> 
> This effectively controls writes onto transaction boundaries. Naively commit()ing a common session would otherwise accidentally write incomplete graphs, and for similar reasons, commit()ing directly via the qtx() would not work.
> 
> I considered that certain in flight modifications may be being committed by the qtx during lazy-load resolution, but since it's autobegin and I never call commit() I'm thinking that's avoided. Likewise, there may be some lazy loads that are being resolved while certain objects are in a write session, but since their sub-graph is coincident, that's ok.
> 

all of the SQLAlchemy ORM's assumptions assume that sessions and ORM mapped objects are never used among multiple threads or greenlets.   if an application does that, it would need to mutex operations at a coarse-grained level, that of app-level operations, not at the lazy loading level.    There's a way above to try to get the mutex in at the attribute access level but IMO this is a really bad idea and if you have problems, I don't have the resources to help with patterns we dont support.

**Reply by aathan** [reply]
I had indeed run across [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=do_orm_execute#sqlalchemy.orm.SessionEvents.do_orm_execute) but what wasn't clear is how to recognize lazy loads vs general query activity. Is there an easy way? Might not need it, but wondering.

> you have to [merge](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=merge#sqlalchemy.orm.Session.merge) those objects into a local session.
> all of the SQLAlchemy ORM's assumptions assume that sessions and ORM mapped objects are never used among multiple threads or greenlets

Clearly you mean never used concurrently or without care. So the question becomes where exactly that line is drawn. Thread safety is not discussed at the above link or within the longer discussion at [Merging](https://docs.sqlalchemy.org/en/14/orm/session_state_management.html#unitofwork-merging).

If we have threadB and threadA and associated seshB and seshA and some ORM oA attached to seshA we could:

(1) Call seshB.merge(oA) from threadA, no mutex
(2) Call seshB.merge(oA) from threadB, no mutex
(3) Either 1 or 2 so long as seshB and seshA are mutexed during the operation
(4) Either 1 or 2 so long as oA is mutexed during the operation
(5) Either 1 or 2 so long as oA and any of its relations are mutexed during the operation
(6) Other things like expunge(oA) first
(7) None of the above, always do this into seshZ, that's used by no one else, and pass seshZ across to threadB

Maybe there's a global statement somewhere that says absolutely no ORM or Session should have any attribute access or API calls made without mutexing all other threads.

Per your statement I should conclude only 3 is safe, but the docs seem to leave room for all of 2-6.  If 3 is the only right answer, a note may be warranted on this API doc.

Obviously a coarse grained mutex methodology, as you suggested, is likely to work. That said, it feels like we're actually quite close to making it work in a more fine grained way too. You're probably aware of a ton of other gotchas so I'll tread lightly.

A.

**Reply by zzzeek** [reply]
> I had indeed run across [do_orm_execute](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=do_orm_execute#sqlalchemy.orm.SessionEvents.do_orm_execute) but what wasn't clear is how to recognize lazy loads vs general query activity. Is there an easy way? Might not need it, but wondering.

that's the easy way, pretty easy.    not really any other way otherwise...

> (1) Call seshB.merge(oA) from threadA, no mutex (2) Call seshB.merge(oA) from threadB, no mutex (3) Either 1 or 2 so long as seshB and seshA are mutexed during the operation (4) Either 1 or 2 so long as oA is mutexed during the operation (5) Either 1 or 2 so long as oA and any of its relations are mutexed during the operation (6) Other things like expunge(oA) first (7) None of the above, always do this into seshZ, that's used by no one else, and pass seshZ across to threadB

the one other thing with this pattern is that the objects you are merging are detached.  so they are safe to read in a concurrent fashion as required by merge().

> 
> Maybe there's a global statement somewhere that says absolutely no ORM or Session should have any attribute access or API calls made without mutexing all other threads.

[is the session threadsafe?](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#is-the-session-thread-safe) - thats the statement.


> 
> Per your statement I should conclude only 3 is safe, but the docs seem to leave room for all of 2-6. If 3 is the only right answer, a note may be warranted on this API doc.

detached objects have no connection to a session so you can do whatever you want with those.   sessions as you noted can start querying SQL at any time and they also have mutable state, so multiple threads acting upon the same session can easily corrupt the state of the Session.


> 
> Obviously a coarse grained mutex methodology, as you suggested, is likely to work. That said, it feels like we're actually quite close to making it work in a more fine grained way too. You're probably aware of a ton of other gotchas so I'll tread lightly.

using a coarse grained locking pattern is just a standard approach, for an application that's organized into some kind of DAO layer, similar, but if you treated your "global objects" as a detached, second level cache then you wouldnt have to mutex anything.   see pattern such as [dogpile.caching](https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.dogpile_caching).

> 
> A.

**Comment by aathan** [comment]
> that's the statement.
Yes! Very clear indeed.

I have understood that merge(obj) called from a thread must either be mutexed on obj's session or obj must be detached.

Quoting from the doc the wording that lead me to believe obj may not need to be detached in the general case:

> The load=False use case adds the caveat that the given object has to be in a “clean” state, that is, has no pending changes to be flushed - even if the incoming object is detached from any [Session](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=merge#sqlalchemy.orm.Session).

Thanks again,
A.
